nameko/rpc.py

```diff
@@ -253,149 +318,214 @@ def setup(self):
+        self.consumer.stop()
+
+    def check_for_lost_replies(self):
+        if self.pending:
+            try:
+                with get_connection(self.amqp_uri) as conn:
+                    self.queue.bind(conn).queue_declare(passive=True)
+            except NotFound:
+                raise ReplyQueueExpiredWithPendingReplies(
+                    "Lost replies for correlation ids:\n{}".format(
+                        "\n".join(self.pending.keys())
+                    )
+                )
 
-    def get_reply_event(self, correlation_id):
+    def register_for_reply(self, correlation_id):
+        """ Register an RPC call with the given `correlation_id` for a reply.
+
+        Returns a function that can be used to retrieve the reply, blocking
+        until it has been received.
+        """
         reply_event = Event()
-        self._reply_events[correlation_id] = reply_event
-        return reply_event
+        self.pending[correlation_id] = reply_event
+        return reply_event.wait
 
     def handle_message(self, body, message):
-        self.queue_consumer.ack_message(message)
+        self.consumer.ack_message(message)
 
         correlation_id = message.properties.get('correlation_id')
-        client_event = self._reply_events.pop(correlation_id, None)
-        if client_event is not None:
-            client_event.send(body)
+        rpc_call = self.pending.pop(correlation_id, None)
+        if rpc_call is not None:
+            rpc_call.send(body)
         else:
             _log.debug("Unknown correlation id: %s", correlation_id)
 
 
-class RpcProxy(DependencyProvider):
+class RpcProxy(DependencyProvider, HeaderEncoder):
+    """ DependencyProvider for injecting an RPC proxy into a service.
+
+    :Parameters:
+        target_service : str
+            Target service name
+        **publisher_options
+            Options to configure the :class:`~nameko.amqqp.publish.Publisher`
+            that sends the message.
+    """
 
-    rpc_reply_listener = ReplyListener()
+    publisher_cls = Publisher
+
+    reply_listener = ReplyListener()
 
-    def __init__(self, target_service, **options):
+    def __init__(self, target_service, **publisher_options):
         self.target_service = target_service
-        self.options = options
+        for option in RESTRICTED_PUBLISHER_OPTIONS:
+            publisher_options.pop(option, None)
+        self.publisher_options = publisher_options
 
-    def get_dependency(self, worker_ctx):
-        return ServiceProxy(
-            worker_ctx,
-            self.target_service,
-            self.rpc_reply_listener,
-            **self.options
-        )
+    @property
+    def amqp_uri(self):
+        return self.container.config[AMQP_URI_CONFIG_KEY]
+
+    def setup(self):
 
+        # ReplyListener.setup() runs concurrently in another thread, so
+        # we need to wait here until it's defined its queue
+        while self.reply_listener.queue is None:
+            time.sleep(.1)
```

Comment:
Workaround until https://github.com/nameko/nameko/wiki/Roadmap#refactor-sub-extension-management is done